geochem_inverse_optimize.py

```diff
@@ -75,56 +75,148 @@ def get_sample_graphs(data_dir: str) -> Tuple[nx.DiGraph, Any]:
     return sample_network, sample_adjacency
 
 
-def reset_sample_network(sample_network: nx.DiGraph) -> None:
-    for _, data in sample_network.nodes(data=True):
-        data["data"].my_value = 0.0
-        data["data"].my_flux = 0.0
-        data["data"].total_flux = 0.0
-
-
-def get_primary_terms(sample_network: nx.DiGraph, obs_data: ElementData) -> List[Any]:
-    # Build the main objective
-    # Use a topological sort to ensure an upstream-to-downstream traversal
-    primary_terms = []
-    for sample_name, my_data in nx_topological_sort_with_data(sample_network):
-        # Set up a CVXPY parameter for each element for each node
-        my_data.my_value = cp.Variable(pos=True)
-
-        # area weighted contribution from this node
-        # TODO(rbarnes): Try scaling area of a region by dividing by total area of all regions. `my_data.normalized_area`
-        my_data.my_flux = my_data.area * my_data.my_value
-
-        # Add the flux I generate to the total flux passing through me
-        my_data.total_flux += my_data.my_flux
-
-        # Normalise observation by mean
-        obs_mean = np.mean(list(obs_data.values()))
-        observed = obs_data[my_data.data.name] / obs_mean
-        normalised_concentration = my_data.total_flux / my_data.total_area
-        primary_terms.append(cp_log_ratio_norm(normalised_concentration, observed))
-
-        if ds := nx_get_downstream(sample_network, sample_name):
-            downstream_data = sample_network.nodes[ds]["data"]
-            # Add our flux to the downstream node's
-            downstream_data.total_flux += my_data.total_flux
-
-    return primary_terms
+class SampleNetwork:
+    def __init__(self, sample_network: nx.DiGraph, sample_adjacency: nx.DiGraph) -> None:
+        self.sample_network = sample_network
+        self.sample_adjacency = sample_adjacency
+        self._site_to_parameter: Dict[str, cp.Parameter] = {}
+        self._primary_terms = []
+        self._regularizer_terms = []
+        self._regularizer_strength = cp.Parameter(nonneg=True)
+        self._problem = None
+        self._build_primary_terms()
+        self._build_regularizer_terms()
+        self._build_problem()
+
+    def _build_primary_terms(self) -> None:
+        for _, data in self.sample_network.nodes(data=True):
+            data["data"].total_flux = 0.0
+
+        # Build the main objective
+        # Use a topological sort to ensure an upstream-to-downstream traversal
+        for sample_name, my_data in nx_topological_sort_with_data(self.sample_network):
+            # Set up a CVXPY parameter for each element for each node
+            my_data.my_value = cp.Variable(pos=True)
+
+            # area weighted contribution from this node
+            # TODO(rbarnes): Try scaling area of a region by dividing by total area of all regions. `my_data.normalized_area`
```

Comment:
Have we done this?

Reply:
No, but don't think its required as mean normalisation of each element seemed fixed the inaccurate solves